domain/application/state/application_endpoint.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"errors"
+	"maps"
+	"slices"
+	"strings"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+
+	coreapplication "github.com/juju/juju/core/application"
+	corecharm "github.com/juju/juju/core/charm"
+	"github.com/juju/juju/core/network"
+	corerelation "github.com/juju/juju/core/relation"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	internalerrors "github.com/juju/juju/internal/errors"
+)
+
+// insertApplicationEndpointsParams contains parameters required to insert
+// application endpoints into the database.
+type insertApplicationEndpointsParams struct {
+	appID     coreapplication.ID
+	charmUUID corecharm.ID
+
+	// EndpointBindings is a map to bind application endpoint by name to a
+	// specific space. The default space is referenced by an empty key, if any.
+	bindings map[string]network.SpaceName
+}
+
+// insertApplicationEndpoints inserts application endpoints into the
+// database based on provided parameters.
+// It validates charm relations, space bindings, and ensures proper endpoint
+// insertion within a transaction.
+func (st *State) insertApplicationEndpoints(ctx context.Context, tx *sqlair.TX, params insertApplicationEndpointsParams) error {
```

Comment:
todo: Update comment to be clear that ALL relations and extra-bindings will be added as `application_endpoints` and `application_extra_endpoints`. Where a binding was provided in params.bindings, the space will be set. I had to dig into the code to find that. Almost requested changed.